### PR TITLE
solution 1: binary_to_decimal_conversion

### PR DESCRIPTION
**Pull request by Alemayehu-Desta**
#### Pull Request Description  
This pull request introduces the following:  

1. **Function Implementation**  
   - `binary_to_decimal`: Converts a binary string to its decimal equivalent.  
   - Ensures input validation by checking for:  
     - Non-empty strings.  
     - Valid binary characters ('0' and '1').  
   - Raises appropriate exceptions (`ValueError`) for invalid inputs.  

2. **Unit Tests**  
   - Comprehensive test coverage using `unittest`.  
   - Test cases include:  
     - Valid binary strings.  
     - Edge cases such as empty strings and invalid characters.  
     - Type validation for non-string inputs.  

---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/binary_to_decimal.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        >>> binary_to_decimal("")  # Raises ValueError: Input must not be empty.
+        Traceback (most recent call last):
+        ...
+        ValueError: Input must not be empty.
+        >>> binary_to_decimal("102")  # Raises ValueError: Input must only contain '0' and '1'.
+        Traceback (most recent call last):
+        ...
+        ValueError: Input must only contain '0' and '1'.
+    """
+    if not isinstance(binary_str, str):
```
Comment (olumide-AI) [inline solutions/binary_to_decimal.py:34]: You can use more inline comments in your code to explain your lines of code 
In line 40 you can put something like this 
# Convert the binary string to its decimal equivalent using int with base 2

**Comment by suhrobmuboraksho** [comment]
Behavior
Files
[yes ] The file name describes the function's behavior
[yes] There is a module docstring in the function file
[yes] The test file's name matches the function file name -
/tests/test_file_name.py
[yes] There is a module docstring in the tests file
Unit Tests
[yes] The test class has a helpful name in PascalCase
[yes] The test class has a docstring
Every unit test has
[yes] A helpful name
[yes] A clear docstring
[yes] Only one assertion
[yes] There is no logic in the unit test
[yes] All tests pass
[yes] There are tests for defensive assertions
[yes] There are tests for boundary cases
Function Docstring
[yes] The function's behavior is described
The function's arguments are described:
[yes] Type
[yes] Purpose
[yes] Other assumptions (eg. if it's a number, what's the expected range?)
The return value is described
[yes] Type
[yes] Other assumptions are documented
The defensive assertions are documented using Raises:
[yes] Each assumption about an argument is checked with an assertion
[yes] Each assertion checks for only one assumption about the argument
[yes] Include 3 or more (passing!) doctests
The Function
[yes] The function's name describes it's behavior
[yes] The function's name matches the file name
[yes] The function has correct type annotations
Strategy
Do's
[yes] Variable names help to understand the strategy
[yes] Any comments are clear and describe the strategy
[yes] Lines of code are spaced to help show different stages of the strategy
Don'ts
[pass] The function's strategy is not described in the documentation
[pass] Comments explain the strategy, not the implementation
[pass] The function does not have more comments than code
If it does, consider finding a new strategy or a simpler implementation
Implementation
[yes] The code passes the formatting checks
[yes] The code passes all Ruff linting checks
[yes] The code has no (reasonable) Pylint errors
In code review, you can decide when fixing a Pylint error is helpful and
when it's too restricting.
[yes] Variables are named with snake_case
[yes] Variable names are clear and helpful
[yes] The code follows the strategy as simply as possible
[yes] The implementation is as simple as possible given the strategy
[yes] The code includes defensive assertions
[yes] Defensive assertions include as little logic as possible

**Comment by suhrobmuboraksho** [comment]
Module docstrings are present for both files.

**Review by suhrobmuboraksho — COMMENTED** [review]
Left some comments.

**Review by olumide-AI — APPROVED** [review]
We have talked through code on a google meet

**Review by suhrobmuboraksho — APPROVED** [review]
Docstring is good.
Unittests are comprehensive.
Edge cases are taken into consideration.
Ready for approval.
